js/perf/viewer/src/main.js

```diff
@@ -0,0 +1,192 @@
+  }
+
+  const dsSpec = DatasetSpec.parse(params.ds);
+  const [perfData, gitRevs] = await getPerfHistory(dsSpec.dataset());
+
+  chartDatasets = new Map();
+  chartLabels = gitRevs.map(rev => rev.slice(0, 6));
+
+  // Gather all test names up ahead of time, since some commits may be missing results.
+  const firstReps = await Promise.all(perfData.map(pd => {
+    invariant(pd.reps instanceof List);
+    return pd.reps.get(0);
+  }));
+  const testNamesSet = new Set();
+  for (const fr of firstReps) {
+    (await keys(fr)).forEach(testName => testNamesSet.add(testName));
+  }
+  // $FlowIssue: thinks Set is incompatible with spread operand... but it is.
+  const testNames = [...testNamesSet];
+
+  const getElapsed = async (testName: string, pd: Struct) => {
+    invariant(pd.reps instanceof List);
+    const reps = await pd.reps.toJS();
+    const elapsedOrNulls = await Promise.all(reps.map(rep => {
+      invariant(rep instanceof NomsMap);
+      // Note: despite how this code is structured, either all reps should have test data for this
+      // value, or none should. Ideally we'd be able to bail at this point.
+      return rep.get(testName).then(d => d !== null && d !== undefined ? d.elapsed / 1e9 : null);
+    }));
+    return elapsedOrNulls[0] !== null ? median(elapsedOrNulls) : null;
+  };
+
+  const getChartData = (testName: string) =>
+    Promise.all(perfData.map(pd => getElapsed(testName, pd)));
+
+  const testChartData = await Promise.all(testNames.map(getChartData));
+  for (let i = 0; i < testNames.length; i++) {
+    chartDatasets.set(testNames[i], testChartData[i]);
+  }
+
+  render();
+}
+
+// Returns the history of perf data with their git revisions, from oldest to newest.
+async function getPerfHistory(ds: Dataset): Promise<[Struct[], string[]]> {
+  const perfData = [], gitRevs = [];
+
+  for (let head = await ds.head(), i = 0; head && i < MAX_PERF_HISTORY; i++) {
+    const val = head.value;
+    invariant(val instanceof Struct);
+    perfData.push(val);
+    gitRevs.push(val.nomsRevision);
+
+    const parentRef = await head.parents.first(); // TODO: how to deal with multiple parents?
+    head = parentRef ? await parentRef.targetValue(ds.database) : null;
+  }
+
+  return [perfData, gitRevs];
+}
+
+// Returns a map of URL param key to value.
+function getParams(): {[key: string]: string} {
+  // Note: this way anything after the # will end up in `params`, which is what we want.
+  const params = {};
+  const paramsIdx = location.href.indexOf('?');
+  if (paramsIdx > -1) {
+    decodeURIComponent(location.href.slice(paramsIdx + 1)).split('&').forEach(pair => {
+      const [k, v] = pair.split('=');
+      params[k] = v;
+    });
+  }
+  return params;
+}
+
+async function render() {
+  if (!chartDatasets) {
+    return;
+  }
+
+  const datasets = [];
+  for (const [testName, elapsed] of chartDatasets) {
+    const [backgroundColor, borderColor] = await genLightAndDarkColors(testName);
```

Comment:
I see. Every test gets a unique color. Didn't we have a PRG used in testing? Could we use that instead?